src/service/viewer-impl.js

```diff
@@ -766,16 +759,15 @@ export class Viewer {
     if (eventType == 'viewport') {
       if (data['paddingTop'] !== undefined) {
         this.paddingTop_ = data['paddingTop'];
-        this.viewportObservable_.fire(
-          /** @type {!JSONType} */ (data));
+        this.messageObservables_[eventType].fire(
+            /** @type {!JSONType} */ (data));
         return Promise.resolve();
       }
       return undefined;
     }
     if (eventType == 'historyPopped') {
-      this.historyPoppedObservable_.fire({
-        newStackIndex: data['newStackIndex'],
-      });
+      this.messageObservables_[eventType].fire(
```

Comment:
We don't want to do this for each separate event type. How about instead we first process few known event types that viewer itself understands and then it simply delegates to any registered listener.

E.g.
```
if (eventType == 'visibilitychange') {
  // process visibility change
  return ...;
}
if (eventType == 'broadcast') {
  // Do broadcast work.
  return ...;
}

// Otherwise, send to the destination
const obserable = this.messageObservables_[eventType];
if (observable) {
  observable.fire(new ViewerMessage(eventType, data));
  return Promise.resolve();
}
```